File: src/main/java/com/stolser/javatraining/webproject/service/impl/UserServiceImpl.java
package com.stolser.javatraining.webproject.service.impl;

import com.stolser.javatraining.webproject.dao.DaoFactory;
import com.stolser.javatraining.webproject.dao.RoleDao;
import com.stolser.javatraining.webproject.model.entity.user.Credential;
import com.stolser.javatraining.webproject.model.entity.user.User;
import com.stolser.javatraining.webproject.connection_pool.ConnectionPool;
import com.stolser.javatraining.webproject.connection_pool.ConnectionPoolProvider;
import com.stolser.javatraining.webproject.dao.exception.StorageException;
import com.stolser.javatraining.webproject.service.UserService;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;

public class UserServiceImpl implements UserService {
    private DaoFactory factory = DaoFactory.getMysqlDaoFactory();
    private ConnectionPool connectionPool = ConnectionPoolProvider.getPool();

    private UserServiceImpl() {
    }

    private static class InstanceHolder {
        private static final UserServiceImpl INSTANCE = new UserServiceImpl();
    }

    /**
     * @return a singleton object of this type
     */
    public static UserService getInstance() {
        return InstanceHolder.INSTANCE;
    }

    @Override
    public User findOneById(long id) {
        try (Connection conn = connectionPool.getConnection()) {

            User user = factory.getUserDao(conn).findOneById(id);

            if (user != null) {
                user.setRoles(factory.getRoleDao(conn).findRolesByUserName(user.getUserName()));
            }

            return user;

        } catch (SQLException e) {
            throw new StorageException(e);
        }
    }

    @Override
    public Credential findOneCredentialByUserName(String userName) {
        try (Connection conn = connectionPool.getConnection()) {

            return factory.getCredentialDao(conn).findCredentialByUserName(userName);

        } catch (SQLException e) {
            throw new StorageException(e);
        }
    }

    @Override
    public User findOneUserByUserName(String userName) {
        try (Connection conn = connectionPool.getConnection()) {
            User user = factory.getUserDao(conn).findOneByUserName(userName);

            if (user != null) {
                user.setRoles(factory.getRoleDao(conn).findRolesByUserName(userName));
            }

            return user;
        } catch (SQLException e) {
            throw new StorageException(e);
        }
    }

    @Override
    public List<User> findAll() {
        try (Connection conn = connectionPool.getConnection()) {
            RoleDao roleDao = factory.getRoleDao(conn);

            List<User> allUser = factory.getUserDao(conn).findAll();

            allUser.forEach(user -> user.setRoles(roleDao.findRolesByUserName(user.getUserName())));

            return allUser;

        } catch (SQLException e) {
            throw new StorageException(e);
        }
    }

    @Override
    public void createNewUser(User user, Credential credential, String userRole) {
        Connection conn = connectionPool.getConnection();

        try {
            conn.setAutoCommit(false);

            long userId = factory.getUserDao(conn).createNew(user);

//            factory.getRoleDao(conn).addRole(userId, userRole);
//            factory.getCredentialDao(conn).createNew(credential);


            conn.commit();
            conn.setAutoCommit(true);

        } catch (Exception e) {
            try {
                conn.rollback();

            } catch (SQLException e1) {
                throw new StorageException(e);
            }

            throw new StorageException(e);
        } finally {
            if (conn != null) {
                try {
                    conn.close();

                } catch (SQLException e) {
                    throw new StorageException(e);
                }
            }
        }
    }


}

File: src/main/scala/com/stolser/javatraining/webproject/service/impl/UserServiceImpl.scala
package com.stolser.javatraining.webproject.service.impl

import com.stolser.javatraining.webproject.dao.AbstractConnection
import com.stolser.javatraining.webproject.model.entity.user.{Credential, User, UserRole}
import com.stolser.javatraining.webproject.service.ServiceUtils.withConnection
import com.stolser.javatraining.webproject.service.UserService

/**
	* Created by Oleg Stoliarov on 11/21/18.
	*/
abstract class UserServiceImpl extends UserService {
	this: ServiceDependency =>

	override def findOneById(id: Long): Option[User] =
		findOneUserBy(UserParameterName.ID, id)

	override def findOneByName(userName: String): Option[User] =
		findOneUserBy(UserParameterName.NAME, userName)

	private object UserParameterName extends Enumeration {
		val ID, NAME = Value
	}

	private def findOneUserBy(paramName: UserParameterName.Value,
														paramValue: Any): Option[User] = {
		withConnection { conn =>
			val userInDb = paramName match {
				case UserParameterName.ID => daoFactory.userDao(conn)
					.findOneById(paramValue.asInstanceOf[Long])
				case UserParameterName.NAME => daoFactory.userDao(conn)
					.findOneByUserName(paramValue.asInstanceOf[String])
			}

			userInDb match {
				case Some(user) =>
					setUserRoles(user, conn)
					Some(user)
				case _ => None
			}
		}
	}

	private def setUserRoles(user: User,
													 conn: AbstractConnection): Unit =
		user.roles = daoFactory.roleDao(conn).findRolesByUserName(user.userName)

	override def findOneCredentialByUserName(userName: String): Option[Credential] =
		withConnection { conn =>
			daoFactory.credentialDao(conn)
				.findCredentialByUserName(userName)
		}

	override def findAll: List[User] =
		withConnection { conn =>
			val roleDao = daoFactory.roleDao(conn)

			for (user <- daoFactory.userDao(conn).findAll)
				yield {
					user.roles = roleDao.findRolesByUserName(user.userName)
					user
				}
		}

	override def createNewUser(user: User,
														 credential: Credential,
														 userRole: UserRole.Value): Boolean = {
		require(user != null)
		require(credential != null)
		require(userRole != null)

		withConnection { conn =>
			conn.beginTransaction()

			val userId = daoFactory.userDao(conn).createNew(user)
			credential.userId = userId
			val isNewCredentialCreated = daoFactory.credentialDao(conn)
				.createNew(credential)

			if (!isNewCredentialCreated) {
				conn.rollbackTransaction()
				return false
			}

			daoFactory.roleDao(conn).addRole(userId, userRole)
			conn.commitTransaction()

			return true
		}
	}

	override def emailExistsInDb(email: String): Boolean =
		withConnection { conn =>
			daoFactory.userDao(conn).emailExistsInDb(email)
		}
}
